usecases/decision_usecase.go

```diff
@@ -217,55 +224,223 @@ func (usecase *DecisionUsecase) CreateDecision(
+		}
+
+		ctx, cancel := context.WithTimeout(ctx, models.DECISION_TIMEOUT)
+		defer cancel()
+		ctx, span := tracer.Start(
+			ctx,
+			"DecisionUsecase.CreateAllDecisions",
+			trace.WithAttributes(attribute.String("scenario_id", scenario.Id)),
+		)
+		defer span.End()
+		scenarioExecution, err := evaluate_scenario.EvalScenario(ctx, evaluationParameters, evaluationRepositories)
+		if errors.Is(err, models.ErrScenarioTriggerConditionAndTriggerObjectMismatch) {
+			nbSkipped++
+			continue
+		} else if err != nil {
+			return nil, 0, errors.Wrap(err, "error evaluating scenario in CreateAllDecisions")
+		}
+
+		decision := models.AdaptScenarExecToDecision(scenarioExecution, payload)
+		items = append(items, decisionAndScenario{decision: decision, scenario: scenario})
+
+	}
+
+	ctx, span2 := tracer.Start(ctx, "DecisionUsecase.CreateAllDecisions - store decisions")
+	defer span2.End()
+	decisions, err = executor_factory.TransactionReturnValue(ctx, usecase.transactionFactory, func(
+		tx repositories.Executor,
+	) ([]models.DecisionWithRuleExecutions, error) {
+		var ids []string
+		for _, item := range items {
+			ids = append(ids, item.decision.DecisionId)
+			if err = usecase.decisionRepository.StoreDecision(
+				ctx,
+				tx,
+				item.decision,
+				input.OrganizationId,
+				item.decision.DecisionId,
+			); err != nil {
+				return nil, fmt.Errorf("error storing decision in CreateAllDecisions: %w", err)
+			}
+
+			if err := usecase.createCaseIfApplicable(ctx, tx, item.scenario,
+				item.decision, input.OrganizationId); err != nil {
+				return nil, err
 			}
 		}
 
-		return usecase.decisionRepository.DecisionById(ctx, tx, newDecisionId)
+		return usecase.decisionRepository.DecisionsByIds(ctx, tx, ids)
 	})
+	return
+}
+
+func (usecase *DecisionUsecase) createCaseIfApplicable(
+	ctx context.Context,
+	tx repositories.Executor,
+	scenario models.Scenario,
+	decision models.DecisionWithRuleExecutions,
+	organizationId string,
+) error {
+	if scenario.DecisionToCaseOutcomes != nil &&
+		slices.Contains(scenario.DecisionToCaseOutcomes, decision.Outcome) &&
+		scenario.DecisionToCaseInboxId != nil {
+		_, err := usecase.caseCreator.CreateCaseAsWorkflow(ctx, tx, models.CreateCaseAttributes{
+			DecisionIds: []string{decision.DecisionId},
+			InboxId:     *scenario.DecisionToCaseInboxId,
+			Name: fmt.Sprintf(
+				"Case for %s: %s",
+				scenario.TriggerObjectType,
+				decision.ClientObject.Data["object_id"],
+			),
+			OrganizationId: organizationId,
+		})
+		if err != nil {
+			return errors.Wrap(err, "error linking decision to case")
+		}
+	}
+	return nil
+}
+
+// used in different contexts, so allow different cases of input: pass client object or raw payload
+func (usecase DecisionUsecase) validatePayload(
+	ctx context.Context,
+	organizationId string,
+	triggerObjectTable string,
+	clientObject *models.ClientObject,
+	rawPayload json.RawMessage,
+) (payload models.ClientObject, dataModel models.DataModel, err error) {
```

Comment:
I still feel unconfortable with that syntax.

I was reading those lines and bugged on "what is the actual return signature of this func"
```go
	if clientObject != nil {
		payload = *clientObject
		return
	}
```

Reply:
Yes I understand, I'm also mostly not a fan.
Still, I have two cases at least where I find it acceptable and possibly preferable:
- case 1 is here: we have 3+ return values, with a "long" zero value. Go code with i/o by construction tends to be cluttered with `return` statements, and while I'm fine with repeating `return nil, err` or `return err` every 4 lines, it gets really tedious to repeat `return models.MyFirstLongModelName{}, models.MySecondLongModelName{}, err` everywhere. Here we can avoid that by assigning `err` and using a naked return. That said, both syntaxes are compatible, perhaps there is a reasonable tradeoff in using full returns for the "good path" return statements.
- case 2 is above: when a return value (especially where there are several return values) is ambiguous in its meaning. Like above, we return `decisions []models.Decision, nbSkipped int, err error`, which is super clear to read, while `[]models.Decision, int, error` is kind of mysterious (ok, we can work with comments but it's less good IMO)